tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+      SmallVector<Value, 4> stride_values(result_rank, one);
+      SmallVector<Value, 4> begin_values(result_rank, zero);
+      begin_values[0] = builder.create<arith::ConstantIndexOp>(loc, i);
+      SmallVector<Value, 4> end_values;
+      end_values.push_back(builder.create<arith::ConstantIndexOp>(loc, i + 1));
+      for (int64_t j = 1; j < result_rank; j++) {
+        end_values.push_back(
+            builder.create<tensor::DimOp>(loc, batched_dot, j));
+      }
+      auto index_ty = builder.getIndexType();
+      auto start_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(begin_values.size())},
+                                index_ty),
+          begin_values);
+      auto end_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(end_values.size())},
+                                index_ty),
+          end_values);
+      auto stride_indices = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get({static_cast<int64_t>(stride_values.size())},
+                                index_ty),
+          stride_values);
+      SmallVector<int64_t, 4> slice_shapes(result_rank,
+                                           ShapedType::kDynamicSize);
+      slice_shapes[0] = 1;
+      for (int64_t i = 1; i < result_rank; i++) {
+        slice_shapes[i] = orig_dot_type.getDimSize(i - 1);
+      }
+      auto slice_type =
+          RankedTensorType::get(slice_shapes, orig_dot_type.getElementType());
+      auto dyn_slice = builder.create<mhlo::RealDynamicSliceOp>(
+          loc, slice_type, batched_dot, start_indices, end_indices,
+          stride_indices);
+      SmallVector<Value, 4> reshape_shape_values;
+      for (int64_t j = 1; j < slice_type.getRank(); j++) {
+        reshape_shape_values.push_back(
+            builder.create<tensor::DimOp>(loc, dyn_slice, j));
+      }
+      auto reshape_shape = builder.create<tensor::FromElementsOp>(
+          loc,
+          RankedTensorType::get(
+              {static_cast<int64_t>(reshape_shape_values.size())}, index_ty),
+          reshape_shape_values);
+      auto dyn_reshape = builder.create<mhlo::DynamicReshapeOp>(
+          loc, orig_dot_type, dyn_slice, reshape_shape);
+      op->replaceAllUsesWith(dyn_reshape);
+    }
+  }
+
+  // No longer need the original dot ops.
+  for (int64_t i = 0; i < ops.size(); i++) {
+    ops[i]->erase();
+  }
+
+  return true;
+}
+
+bool DotBatchingConverter::batchingDots() {
+  // Dot ops in `equal_shape_dot_map_` can be batched together if the batching
+  // does not introduce cycle.
+
+  // Form cycle detector.
+  std::vector<Operation*> func_op_list;
+  DenseMap<Operation*, int64_t> func_op_to_id;
+  func_.walk([&](Operation* op) {
+    func_op_to_id.try_emplace(op, func_op_list.size());
+    func_op_list.push_back(op);
+  });
+  GraphCycles cycle_detector(func_op_list.size());
```

Comment:
cycle detector should only be used at basic block level. if we want to fuse dot general across basic block, then we should take region-based control flow into consideration.

BTW, dot batching may be viewed as dot fusion as well. If so, we could re-use  a log of code in the fusion pass.  The bad news is that we do fusion on buffer level a.t.m. We may re-visit this in the future.

Reply:
We only batch dots within the same block (see the checking at L500~503). Is it safe? Otherwise it needs to get the block just wraps each dot-op and build cycle-detector one by one.

Reply:
> cycle detector should only be used at basic block level. if we want to fuse dot general across basic block, then we should take region-based control flow into consideration.
> 
> BTW, dot batching may be viewed as dot fusion as well. If so, we could re-use a log of code in the fusion pass. The bad news is that we do fusion on buffer level a.t.m. We may re-visit this in the future.

It is a kind of fusion. Beside the problem that current fusion is for buffer level ops, it do not fully work for remote fusion (i.e., horizontal fusion) even with some code over there. Dot batching is remote fusion. If we plan to enable remote fusion, some common tool-funcs can be reused.

Reply:
I also have similar points in offline discussion that DotMerge is better to be implemented as another FusionKind of the FusionPass. The only problem is that the 'horizontal fusion' is actually not working in the current codebase. I tried it for some other reasons a couple of weeks ago and it seems to me we do need some more fix. I didn't put much effort in investigating the root cause. But if we are clearer about the effort in fixing it, i would tend to choose to do this in the FusionPass instead of a standalone pass.

Reply:
It worths an offline discussion. My current point is that the only common logic here is the cyclic detector. All the other logics are different. By introducing a new FusionKind, it introduces more codes than saving some codes. But I may missed something. Let's discuss it offline.

Reply:
I left a TODO at the begging of the file about the possible plan of merging this pass into fusion-pass when horizontal fusion is fixed.